Black and White listings for the Extension Manager

echarles opened the pull request:
This is a PR to implement Black and White listings for the Extension Manager as discussed in https://github.com/jupyterlab/jupyterlab/issues/7933 and https://github.com/jupyterlab/jupyterlab/issues/7967.

It must be seen in conjunction with https://github.com/jupyterlab/jupyterlab_server/pull/82

echarles commented:
Thx @saulshanabrook for looking at that. I should have bring more context around the work done so far.

First, I have added a comment [here](https://github.com/jupyterlab/jupyterlab/issues/7933#issuecomment-596060306) further to the discussion at the jupyterlab meeting. The idea is to not allow the user to change the URIs via settings but rather to enforce them via configuration on server side.

Therefor, this PR does an initial fetch on thee server to get the URIs and then fetches the effective listings from whatever defined (github repo...)

I have added a simple example in `example/listings` folder for dev/experiment purposes that exposes the listings as an additional handler on the jupyterlab_server.

But yes, the goal is to host those listings on github.

Hope this clarifies a bit. I still need to document all that but I will do it when it will be a bit more stable.

echarles commented:
Current status.

![demo](https://user-images.githubusercontent.com/226720/76164032-ac564d80-614b-11ea-8766-50b98915799a.gif)

saulshanabrook commented on docs/source/user/extensions.rst:
```diff
@@ -194,6 +249,138 @@ will be up to you to take these into account or not.
+~~~~~~~~
+
+When searching extensions, JupyterLab displays the complete search result and 
+the user if free to install any extension. This is the :ref:`default_mode`.
```
```suggestion
the user is free to install any extension. This is the :ref:`default_mode`.
```

saulshanabrook commented on docs/source/user/extensions.rst:
```diff
@@ -194,6 +249,138 @@ will be up to you to take these into account or not.
+^^^^^^^^^^^^
+
+In the ``default`` mode, no listing is enabled and the search behavior is unchanged and
+is the one described previously.
```
```suggestion
In the ``default`` mode, no listing is enabled and the search behavior is unchanged,
as described previously.
```

echarles replied:
Committed. Thx.

saulshanabrook commented on docs/source/user/extensions.rst:
```diff
@@ -194,6 +249,138 @@ will be up to you to take these into account or not.
+
+If you, or your administrator, has enabled the whitelist mode
+JupyterLab will use the whitelist and only show extensions present
+in the withelist. The other extensions will not be show in the search result.
```
```suggestion
in the whitelist. The other extensions will not be show in the search result.
```

echarles replied:
Committed. Thx.

saulshanabrook commented on docs/source/user/extensions.rst:
```diff
@@ -194,6 +249,138 @@ will be up to you to take these into account or not.
+A whitelist maintains a set of approved extensions that users can freely 
+search and install. Extensions need to go through some sort of vetting process 
+before they are added to the whitelist. When using a whitelist, the extension manager 
+will highlight extensions that have been explicitly added to the whitelist.
```
```suggestion
will only show extensions that have been explicitly added to the whitelist.
```

saulshanabrook commented on docs/source/user/extensions.rst:
```diff
@@ -194,6 +249,138 @@ will be up to you to take these into account or not.
+         "name": "@jupyterlab-examples/launcher",
+         "type": "jupyterlab",
+         "reason": "@jupyterlab-examples/launcher is blacklisted for test purpose - Do NOT take this for granted!!!",
+         "creation_date": "2020-03-11T03:28:56.782Z",
```
Thoughts on removing all keys besides `name` or making them optional?

Just wondering what they add to the UI at the moment and wanting to ensure it's as easy as possible to maintain these files.

echarles replied:
Made the name mandatory and the others optional. Also documented that.

saulshanabrook commented on jupyterlab/labapp.py:
```diff
@@ -14,7 +14,7 @@
 from notebook.notebookapp import NotebookApp, aliases, flags
 from notebook.utils import url_path_join as ujoin
-from traitlets import Bool, Instance, Unicode
+from traitlets import Bool, Instance, Unicode, Integer
```
```suggestion
from traitlets import Bool, Instance, Unicode
```

echarles replied:
Committed. Thx.

saulshanabrook commented on packages/extensionmanager/src/listings.ts:
```diff
@@ -0,0 +1,139 @@
+ * Listing search result structure (subset).
+ *
+ */
+export interface IListResult {
```
Think this might be better suited as a type union, since you can only have a blacklist or whitelist, not both. Something like this:

```typescript
type ListResult = null | {mode: 'white' | 'black', uris: string[], entries: IListEntry[] }
```

echarles replied:
Makes sense. Updates pushed with your proposal.

saulshanabrook commented on packages/extensionmanager/src/listings.ts:
```diff
@@ -0,0 +1,139 @@
+}
+
+/**
+ * Call the API extension
```
```suggestion
 * Call the listings API extension
```

echarles replied:
Done. Thx.

saulshanabrook commented on packages/extensionmanager/src/model.ts:
```diff
@@ -708,8 +807,16 @@ export class ListModel extends VDomModel {
   private _debouncedUpdate: Debouncer<void, void>;
+
+  private _listMode: 'black' | 'white' | null;
+  private _blacklistMap: Map<string, IListEntry>;
```
Curious why this is not just an `Array<IListEntry>`. The string key just seems to be the name?

echarles replied:
I was showing before more details in the UI (reasons for blaklistings...). Then we decided to not show that. I propose to leave it as such, I like knowing that it is an array of listentry, not simply an array of string.

saulshanabrook replied:
Yeah, so it could be an `Array<IListEntry>` not an `Array<string>`? I think you have this as an existing todo anyway: https://github.com/jupyterlab/jupyterlab/pull/7989/files/bc7b5d1efe06ae2c2feb5161b56d3f00ce30f70a#diff-4c52b79547ae3ec18aa8f0172bca4f58R542

echarles replied:
yep. my first implementation was not using regexp so a Map was better. Now with the regexp an Array is enough. I will pass on all current and new TODO tomorrow.

echarles replied:
Replacing `tomorrow` with `today` (always better :)) - Map have been replace with Array. Resolving.

echarles commented:
> Thanks Eric! I tried this locally with a custom blacklist and whitelist and it worked for me. Left a few minor comments.

Thx @saulshanabrook I have implemented and commented on your reviews.

> I think a follow up to this could be enabling the extension manager by default and setting a default blacklist to a Jupyter maintained URL.

Sure, now that we have better ways to control security, a logical next step is to enable by default. We can discuss that at today weekly meeting.

I will now focus getting CI green.

echarles commented:
@saulshanabrook For CI, looking at Azure (not at the new GitHub Actions), we have 2 failures:

- linux docs due to `(line  107) broken    ***/pull/7794 - 429 Client Error: too many requests for url: ***/pull/7794` - That failing command runs fine on my local env.
- windows js fails with timeout (linux js runs fine) - I have seen quite some time that and would put the cause on a flaky windows env.

saulshanabrook commented:
> linux docs due to (line 107) broken ***/pull/7794 - 429 Client Error: too many requests for url: ***/pull/7794 - That failing command runs fine on my local env.

Yeah I was getting that as well. I think docs builds tests are broken now b/c Github changed how it was rate limiting it's APIs so our link checker is getting errors.

> windows js fails with timeout (linux js runs fine) - I have seen quite some time that and would put the cause on a flaky windows env.

Yep.

MarsBarLee commented:
I have a suggestion- we could change the checkbox to a red 'Enable' button and a blue 'Disable' button. My reasoning is that a red 'Enable' button acts more as a stronger cautionary warning than the checkbox. By creating more 'friction' in the user's experience, it requires more thought and puts more liability on the user's side, rather than JupyterLab's side.
![Screen Shot 2020-03-25 at 10 59 36 AM](https://user-images.githubusercontent.com/46167686/77585148-68d54080-6eba-11ea-8afd-53d9f4a17c94.png)

After enabling:
![Screen Shot 2020-03-25 at 10 59 48 AM](https://user-images.githubusercontent.com/46167686/77585184-74286c00-6eba-11ea-90c5-1294c2643a57.png)

saulshanabrook commented:
@MarsBarLee I like how it looks better too!

tgeorgeux commented:
@MarsBarLee That looks good. I don't have a strong opinion, but we also discussed grey for the disable button. I have a weak preference for blue over grey. In that once that warning is opened blue is the lowest friction 'disable this dangerous thing' option. I like it a lot better than the check box.

MarsBarLee commented:
Hey Eric, I have my feedback here.

### UX Review
##### For the warning text:
- Change the font size to ```--jp-ui-font-size1```.
- Change font weight to ```regular```.
- Add a period at end of first paragraph. 
- Add a page break in-between paragraphs.(You could also add 8px padding between first and second paragraph if that's easier.)
#### For the UI, when the extension installer is disabled
- The DISCOVER heading should be disabled and greyed out. Use ```--jp-layout-color3``` for the disabled heading. (If it's too diffilcult to add the color change, at least close and disable interactions)
- The INSTALLER header should be be disabled and collapsed as well. Use ``` --jp-layout-color3``` as well. (Same as above)
- Add the ```disabled``` property to the refresh button
#### For the search bar at the top, change the icon to the same icon already used in the command palette search bar.
With these changes included we should be good to include in JupyterLab 2.1

echarles commented:
@MarsBarLee Thx for the enhancements. I have implemented the following:

Warning text:
[x] Change the font size to --jp-ui-font-size1
[x] Change font weight to regular.
[x] Add a period at end of first paragraph.
[x] Add a page break in-between paragraphs

When the extension installer is disabled.
[x] The DISCOVER and INSTALLER heading should be disabled and greyed out.
[x] Add the disabled property to the refresh button
For the search bar at the top
[ ] change the icon to the same icon already used in the command palette search bar => The search widget in command palette is different from the one used in 
the extension manager. I have done CSS work to make them identical but at some point, I face issues which are not solvable unless I completely migrate one solution
to the other solution. I have just changed the icon which even does not display at the same scale. The blue box in the command palette can not be removed easily
(I guess it has been added on purpose...)

Here is what it gives.

![Untitled](https://user-images.githubusercontent.com/226720/77778070-c00d1980-7050-11ea-9d20-60d8e44e9d2d.gif)

saulshanabrook commented:
I am trying this locally without a blacklist or whitelist. It seems like there might be some error there in that logic? The server seems to returning no whitelist of blacklist URLs (as you can see from the network response here), but it is saying an extension is not on the whitelist.

Could you try to verify this locally to see if it works for you?

<img width="1817" alt="Screen Shot 2020-03-27 at 1 57 27 PM" src="https://user-images.githubusercontent.com/1186124/77785676-e7ed8480-7032-11ea-8770-726ea7b0f176.png">

echarles commented:
Gosh, you saved the lab from complete extension breakdown..., Saul. I have pushed https://github.com/jupyterlab/jupyterlab/pull/7989/commits/66e492103402dab05d8b44446f62b4b5f122846f  to fix that and also tested everything in the 3 modes (default, black, white).

echarles commented:
@saulshanabrook Based on our conversation, I have pushed https://github.com/jupyterlab/jupyterlab/pull/7989/commits/de5de536b123bd1f4d2535111bba94e6495a44e8 to ensure that the extensions discovery tab is populated on first manager display (this is a behavior change compared to master, but as discussed it is a good enhancement).

CI is back green (except the notable `Linux Docs`, but we know that we can keep it Green as it is a know issue tracked in https://github.com/jupyterlab/jupyterlab/pull/8068.

LGTM to merge.

saulshanabrook commented on packages/extensionmanager/src/widget.tsx:
```diff
@@ -577,25 +732,30 @@ export class ExtensionView extends VDomRenderer<ListModel> {
+              }}
+              performAction={this.onAction.bind(this)}
+            />
+          </>
```
Need to keep at top level so react doesn't complain about keys not being found. I tried this locally.

```suggestion
          <ListView
            key="search-items"
            listMode={model.listMode}
            viewType={'searchResult'}
            // Filter out installed extensions:
            entries={model.searchResult.filter(
              entry => model.installed.indexOf(entry) === -1
            )}
            numPages={pages}
            onPage={value => {
              this.onPage(value);
            }}
            performAction={this.onAction.bind(this)}
          />
```

echarles replied:
Thx Saul for the fix. Committed your suggestion.
